Make MSVC detection ludicrously robust

retep998 opened the pull request:
Resurrection of https://github.com/rust-lang/rust/pull/31158

r? @alexcrichton

alexcrichton commented on src/librustc_trans/back/msvc/mod.rs:
```diff
@@ -223,31 +256,25 @@ mod platform {
+                _ => vec![],
             }
-        } else { None }
+        } else { vec![] }
```
Perhaps this could just be:

```rust
match (arch, host_arch()) {
    // ... 
}
```

Can you also add docs as to what's being returned here? It's not clear to me at least what these are vectors of nor what each element of the pair is.

alexcrichton commented on src/librustc_trans/back/msvc/mod.rs:
```diff
@@ -42,110 +44,139 @@ mod platform {
+        }).and_then(|((mut cmd, host), sub)| {
+            get_ucrt_dir().map(|dir| {
+                debug!("Found Universal CRT {:?}", dir);
+                add_lib(&mut cmd, &dir.join("ucrt").join(sub));
```
This is pretty confusing using `map` and `and_then` purely for the side effect of calling `add_lib`, can this switch to using a macro or the original form of control flow?

retep998 commented:
Updated `bin_subdir` and the comments for it.

Also added a commit with one function converted to use `otry!`. Can you tell me what you think and whether I should apply that style to the rest of the functions?

alexcrichton commented:
Yeah the logic has gotten complicated enough in a few places that I'd be fine using that macro for most of the functions, other than that looks good to me!

retep998 commented:
I've converted what I could to using the macro.

kennytm commented on src/librustc_trans/back/msvc/mod.rs:
```diff
@@ -31,8 +31,18 @@
+        Some(val) => val,
+        None => return None,
+    })
+}
```
This macro has been written at least twice in `librustc` under different names 😆 

* `try_opt!`: https://github.com/rust-lang/rust/blob/382ab92ceedc258e794c1a95aef21d3be3fa76c4/src/librustc/dep_graph/dep_node.rs#L13-L20
* `option_try!`: https://github.com/rust-lang/rust/blob/bb4b3fb7f97924919f072ec9a360bdf943218dbf/src/librustc/util/common.rs#L72-L74

retep998 replied:
Clearly this just indicates the need for this sort of macro in libstd. Sure would be nice if the new `?` stuff supported `Option`.

alexcrichton commented:
Tidy looks like it's failing on travis, but otherwise r=me

retep998 commented:
Travis appears to have passed.

sfackler commented:
@bors r=alexcrichton

bors commented:
:pushpin: Commit a1b33b4 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton a1b33b4fdc74e0a0b9e48a7f0fd5909d5984979a -->

bors commented:
:hourglass: Testing commit a1b33b4 with merge ed33659...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1626)

alexcrichton commented:
@bors: retry

* https://github.com/rust-lang/rust/issues/34595

alexcrichton commented:
@bors: retry

* v

On Fri, Jul 1, 2016 at 5:38 AM, bors <notifications@github.com> wrote:

> 💔 Test failed - auto-win-msvc-64-opt-rustbuild
> <https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1626>
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rust/pull/34492#issuecomment-229936133>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAD95Laha4Rqk-EAzk9d8w8-n1iLQqWrks5qRQoogaJpZM4I-q8o>
> .
>

bors commented:
:hourglass: Testing commit a1b33b4 with merge 23095ae...

bors commented:
:broken_heart: Test failed - [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1646)

alexcrichton commented:
@bors: retry 

* #34595

bors commented:
:hourglass: Testing commit a1b33b4 with merge 7e07e31...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/9699), [auto-linux-32-opt](https://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/9739), [auto-linux-64-cargotest](https://buildbot.rust-lang.org/builders/auto-linux-64-cargotest/builds/993), [auto-linux-64-cross-freebsd](https://buildbot.rust-lang.org/builders/auto-linux-64-cross-freebsd/builds/804), [auto-linux-64-debug-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-debug-opt/builds/2965), [auto-linux-64-nopt-t](https://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/9676), [auto-linux-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/9657), [auto-linux-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-mir/builds/1250), [auto-linux-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-linux-64-opt-rustbuild/builds/1628), [auto-linux-64-x-android-t](https://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/9651), [auto-linux-cross-opt](https://buildbot.rust-lang.org/builders/auto-linux-cross-opt/builds/2842), [auto-linux-musl-64-opt](https://buildbot.rust-lang.org/builders/auto-linux-musl-64-opt/builds/4807), [auto-mac-32-opt](https://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/9719), [auto-mac-64-opt](https://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/9690), [auto-mac-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-mac-64-opt-rustbuild/builds/1645), [auto-mac-cross-ios-opt](https://buildbot.rust-lang.org/builders/auto-mac-cross-ios-opt/builds/815), [auto-win-gnu-32-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt/builds/4775), [auto-win-gnu-32-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-gnu-32-opt-rustbuild/builds/1648), [auto-win-gnu-64-opt](https://buildbot.rust-lang.org/builders/auto-win-gnu-64-opt/builds/4762), [auto-win-msvc-32-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-32-opt/builds/3971), [auto-win-msvc-64-cargotest](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-cargotest/builds/999), [auto-win-msvc-64-opt](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt/builds/4771), [auto-win-msvc-64-opt-mir](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-mir/builds/1262), [auto-win-msvc-64-opt-rustbuild](https://buildbot.rust-lang.org/builders/auto-win-msvc-64-opt-rustbuild/builds/1653)
